### PR TITLE
Waste Free SMC available for adaptive tempered and tempered SMC. 

### PR DESCRIPTION
**Pull request by ciguaran**
Although SMC base.py was ready for waste-free, tempered and adaptive tempered are coupled to using classic SMC (iterate MCMC and keep the last sample). This PR makes Waste-free available at all levels of the hierarchy.
https://github.com/blackjax-devs/blackjax/issues/710

### CODE REVIEW

File: blackjax/smc/tempered.py [hunk 1]
```diff
@@ -48,12 +48,42 @@ def init(particles: ArrayLikeTree):
...
+def update_and_take_last(
+    mcmc_init_fn,
+    tempered_logposterior_fn,
+    shared_mcmc_step_fn,
+    num_mcmc_steps,
+    n_particles,
+):
+    """
+    Given N particles, runs num_mcmc_steps of a kernel starting at each particle, and
+    returns the last values, waisting the previous num_mcmc_steps-1
```
Comment (junpenglao) [inline blackjax/smc/tempered.py:60]:
```suggestion
    returns the last values, discarding the previous num_mcmc_steps-1
```

**Review by junpenglao — DISMISSED** [review]

**Comment by AdrienCorenflos** [comment]
By the way, whilst we are on that, I remember being a bit dubious about waste free for parallel environments: when you have a lot of processors the analysis in the paper does not hold anymore.
@hai-dang-dau have you ever done some back of the envelope calculations on this? I'd like to be able to recommend people to switch off waste free (and maybe increase the number of intermediary distributions in exchange?) if they have a certain number of processors.

**Comment by hai-dang-dau** [comment]
It's true that waste-free SMC shines best when the number of processors is small relative to the number of particles. (Say, a dozen of processors for several thousand of particles.) I think this is the case for most practical applications.

If your number of particles is comparable to that of processors (say, 500 processors and 500 particles), then yes standard SMC is better.

However, if you have that many processors, you probably have a lot of memory as well. When you increase the number of particles to about a million (still with 500 CPUs), waste-free starts to be preferable again.

**Comment by AdrienCorenflos** [comment]
Well, not really: a commercial grade GPU has fairly little memory so it's
not all that black and white. I'll have a think.

**Comment by hai-dang-dau** [comment]
In my previous post, by "processors" I meant the number of MCMC chains that can be run on parallel. Although GPU has a lot of parallel processing capacity, I'm not sure how this translates into the number of MCMC chains. The delicate point is that one MCMC chain itself involves a lot of parallelizable operations (ie matrix multiplications for calculating the likelihood or for generating preconditioned proposals). It is these primitive operations that are usually parallelized automatically by JAX/XLA, not the higher level operations that were taken into account in my reasoning above.

Having said that, it's true that we developed and tested waste-free SMC with CPUs in mind; so it'd definitely be relevant to have recommendations tailored to GPU-accelerated environment.

File: blackjax/smc/waste_free.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        """
+        states, infos = jax.vmap(mcmc_kernel)(rng_key, position, step_parameters)
+
+        # step particles is num_resmapled, num_mcmc_steps, dimension_of_variable
+        # want to transformed into num_resampled * num_mcmc_steps, dimension of variable
+        def reshape_step_particles(x):
+            if len(x.shape) > 2:
+                return x.reshape((x.shape[0] * x.shape[1], -1))
+            else:
+                return x.flatten()
```
Comment (junpenglao) [inline blackjax/smc/waste_free.py:58]:
```suggestion
        def reshape_step_particles(x):
            num_resmapled, num_mcmc_steps, *dimension_of_variable = x.shape
            return x.reshape((num_resmapled*num_mcmc_steps, *dimension_of_variable))
```

**Review by junpenglao — APPROVED** [review]
Minor nit

**Review by junpenglao — DISMISSED** [review]

**Comment by junpenglao** [comment]
Feel free to submit after fixing the code style

**Comment by ciguaran** [comment]
@junpenglao the review gets dismissed whenever we commit something new, even if it is just github's commiting a reviewer suggestion

**Pull request merged**